protocol/shannon/fullnode_cache.go

```diff
@@ -34,138 +33,136 @@ const (
-	return app, nil
-}
-
-// getFromAppCache retrieves an application from the cache using double-check locking pattern.
-// It returns the cached application and a boolean indicating if it was found.
-func (cfn *cachingFullNode) getFromAppCache(cacheKey string) (*apptypes.Application, bool) {
-	// Check cache first
-	if cachedApp, found := cfn.appCache.Get(cacheKey); found {
-		return cachedApp.(*apptypes.Application), true
-	}
-
-	// Use mutex to prevent multiple concurrent cache updates for the same app
-	cfn.appMutex.Lock()
-	defer cfn.appMutex.Unlock()
-
-	// Double-check cache after acquiring lock
-	if cachedApp, found := cfn.appCache.Get(cacheKey); found {
-		return cachedApp.(*apptypes.Application), true
-	}
-
-	return nil, false
+	// eg. "app:pokt1up7zlytnmvlsuxzpzvlrta95347w322adsxslw"
+	appCacheKey := fmt.Sprintf("%s:%s", appCacheKeyPrefix, appAddr)
+
+	// See: https://github.com/viccon/sturdyc?tab=readme-ov-file#get-or-fetch
+	return cfn.appCache.GetOrFetch(
+		ctx,
+		appCacheKey,
+		func(fetchCtx context.Context) (*apptypes.Application, error) {
+			return cfn.lazyFullNode.GetApp(fetchCtx, appAddr)
+		},
+	)
 }
 
 // GetSession returns the session for the given service and app, using a cached version if available.
+// The cache will automatically refresh the session in the background before it expires.
 func (cfn *cachingFullNode) GetSession(
 	ctx context.Context,
 	serviceID protocol.ServiceID,
 	appAddr string,
 ) (sessiontypes.Session, error) {
-	// Create a unique cache key for this service+app combination
-	sessionCacheKey := createCacheKey(sessionCacheKeyPrefix, fmt.Sprintf("%s:%s", serviceID, appAddr))
-
-	// Try to get session from cache with double-check locking pattern
-	if cachedSession, found := cfn.getFromSessionCache(sessionCacheKey); found {
-		return cachedSession, nil
-	}
-
-	// Cache miss - get from underlying node
-	session, err := cfn.lazyFullNode.GetSession(ctx, serviceID, appAddr)
-	if err != nil {
-		return sessiontypes.Session{}, err
-	}
-
-	// Store in cache if the session is found.
-	cfn.sessionCache.Set(sessionCacheKey, session, cache.DefaultExpiration)
-
-	return session, nil
-}
-
-// getFromSessionCache retrieves a session from the cache using double-check locking pattern.
-// It returns the cached session and a boolean indicating if it was found.
-func (cfn *cachingFullNode) getFromSessionCache(cacheKey string) (sessiontypes.Session, bool) {
-	// Check cache first
-	if cachedSession, found := cfn.sessionCache.Get(cacheKey); found {
-		return cachedSession.(sessiontypes.Session), true
-	}
-
-	// Use mutex to prevent multiple concurrent cache updates for the same session
-	cfn.sessionMutex.Lock()
-	defer cfn.sessionMutex.Unlock()
-
-	// Double-check cache after acquiring lock
-	if cachedSession, found := cfn.sessionCache.Get(cacheKey); found {
-		return cachedSession.(sessiontypes.Session), true
-	}
-
-	return sessiontypes.Session{}, false
+	// eg. "session:eth:pokt1up7zlytnmvlsuxzpzvlrta95347w322adsxslw"
```

Comment:
Move the functionality/logic for sesison key generation to a separate file.

Reply:
I could do a separate function but IMO a separate file is too much.